### PR TITLE
Remove left-linearity checking (deprecates `-nl` option)

### PR DESCRIPTION
**Pull request by francoisthire**
Remove option "-nl" as suggested in #103. Non-linear rules are now allowed by default.

### CODE REVIEW

File: kernel/rule.ml [hunk 1]
```diff
@@ -171,7 +171,7 @@ type pattern_info =
 
 (* ************************************************************************** *)
 
-let allow_non_linear = ref false
+let allow_non_linear = ref true
```
Comment (rlepigre) [inline kernel/rule.ml]: I guess everything should be removed if this option is to disappear... ;)

**Comment by francoisthire** [comment]
Well, I was wondering if it was interesting to let that option in the API.
Since it is only two lines of code, I am not sure if I should keep it or
not.

Le mer. 22 août 2018 à 10:01, Rodolphe Lepigre <notifications@github.com> a
écrit :

**Comment by rlepigre** [comment]
It is not only two lines, there is also the ``is_linear`` function I guess. As I told you yesterday, I think that this functionality is not useful in Dedukti. Of course, a confluence checker (if we ever write one) will possibly reject non-linear rules, but that is not Dedukti's problem.

**Comment by francoisthire** [comment]
The failure cannot be fixed in that PR, because the external tests use an external Makefile.

**Comment by rafoo** [comment]
> a confluence checker (if we ever write one)

What is the problem with the current confluence checker based on CSI^HO (option -cc)?

**Comment by rlepigre** [comment]
> What is the problem with the current confluence checker based on CSI^HO (option -cc)?

There is no problem, but this should be the responsibility of the confluence checker to fail in case of non-left-linear rule. There is nothing that says that there is no confluence criterion compatible with non-left-linear rules if I'm not mistaken.

**Comment by Gaspi** [comment]
I think we should still keep the flag `-nl` (doing nothing) for the next version of Dedukti and remove it completely for the next version when everyone has had time to fix their projects.
Quick note: Travis fails on current PR because some tests expect to be able to pass `-nl` to Dedukti.

**Comment by Gaspi** [comment]
I also believe we should have a new flag `-cl` (check linearity) implementing the previous behavior: non-linear rules are rejected. It's free, the code is already here and maybe some day someone might need it.

**Comment by fblanqui** [comment]
I don't see the point in keeping anything related to linearity. Removing 
-nl will not break any development. Of course, if -nl is removed, every 
test related to -nl should be removed as well. The PR needs to be 
updated accordingly.


Le 23/08/2018 à 09:19, fthire a écrit :

**Pull request merged**